app/src/main/java/org/thoughtcrime/securesms/database/AttachmentDatabase.java

```diff
@@ -0,0 +1,1537 @@
+
+        if (attachments == null) {
+          attachments = new LinkedList<>();
+          output.put(attachment.getMmsId(), attachments);
+        }
+
+        attachments.add(attachment);
+      }
+    }
+
+    return output;
+  }
+
+  public boolean hasAttachment(@NonNull AttachmentId id) {
+    SQLiteDatabase database = databaseHelper.getSignalReadableDatabase();
+
+    try (Cursor cursor = database.query(TABLE_NAME,
+                                        new String[]{ROW_ID, UNIQUE_ID},
+                                        PART_ID_WHERE,
+                                        id.toStrings(),
+                                        null,
+                                        null,
+                                        null)) {
+      if (cursor != null && cursor.getCount() > 0) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  public boolean hasAttachmentFilesForMessage(long mmsId) {
+    String   selection = MMS_ID + " = ? AND (" + DATA + " NOT NULL OR " + TRANSFER_STATE + " != ?)";
+    String[] args      = new String[] { String.valueOf(mmsId), String.valueOf(TRANSFER_PROGRESS_DONE) };
+
+    try (Cursor cursor = databaseHelper.getSignalReadableDatabase().query(TABLE_NAME, null, selection, args, null, null, "1")) {
+      return cursor != null && cursor.moveToFirst();
+    }
+  }
+
+  public @NonNull List<DatabaseAttachment> getPendingAttachments() {
+    final SQLiteDatabase           database    = databaseHelper.getSignalReadableDatabase();
+    final List<DatabaseAttachment> attachments = new LinkedList<>();
+
+    Cursor cursor = null;
+    try {
+      cursor = database.query(TABLE_NAME, PROJECTION, TRANSFER_STATE + " = ?", new String[] {String.valueOf(TRANSFER_PROGRESS_STARTED)}, null, null, null);
+      while (cursor != null && cursor.moveToNext()) {
+        attachments.addAll(getAttachments(cursor));
+      }
+    } finally {
+      if (cursor != null) cursor.close();
+    }
+
+    return attachments;
+  }
+
+  @SuppressWarnings("ResultOfMethodCallIgnored")
+  public void deleteAttachmentsForMessage(long mmsId) {
+    Log.d(TAG, "[deleteAttachmentsForMessage] mmsId: " + mmsId);
+
+    SQLiteDatabase database = databaseHelper.getSignalWritableDatabase();
+    Cursor         cursor   = null;
+
+    try {
+      cursor = database.query(TABLE_NAME, new String[] {DATA, CONTENT_TYPE, ROW_ID, UNIQUE_ID}, MMS_ID + " = ?",
+                              new String[] {mmsId+""}, null, null, null);
+
+      while (cursor != null && cursor.moveToNext()) {
+        deleteAttachmentOnDisk(CursorUtil.requireString(cursor, DATA),
+                               CursorUtil.requireString(cursor, CONTENT_TYPE),
+                               new AttachmentId(CursorUtil.requireLong(cursor, ROW_ID),
+                                                CursorUtil.requireLong(cursor, UNIQUE_ID)));
+      }
+    } finally {
+      if (cursor != null)
+        cursor.close();
+    }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 353:366, 405:418.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.